docs/in-depth-guide-html-scraping.md

```diff
@@ -1,10 +1,180 @@
+
+```python
+from ._utils import normalize_string
+
+# ...
+ingredient_tag = self.soup.find("li", class_="wprm-recipe-ingredient")
+ingredient_text = normalize_string(ingredient_tag.get_text())
+```
+
+### Getting yields
+
+A convenience function called `get_yields()` is provided in the `_utils` package. This function accepts a `str` or `bs4.element.Tag` and will return the yield, handling many of the common formats yields can appear in and normalizing them to a standard format.
+
+```python
+from ._utils import get_yields
+
+# ...
+yield_tag = self.soup.find(class_="wprm-recipe-servings")
+yield_text = get_yields(yield_tag)
+# or
+yield_text = get_yields(yield_tag.get_text())
+# both return '4 servings'
+```
+
+### Getting times
+
+A convenience function called `get_minutes()` is provided in the `_utils` package. This function accepts a `str` or `bs4.element.Tag` and will return the number of minutes as an `int`. This function handles a number of common formats that times can be expressed in.
+
+```python
+from ._utils import get_minutes
+
+# ...
+prep_time_tag = self.soup.find(class_="wprm-recipe-prep_time-minutes")
+prep_time_value = get_minutes(prep_time_tag)
+# or
+prep_time_value = get_minutes(prep_time_tag.get_text())
+# both return 25
+```
+
+## Finding multiple elements
+
+Some information in a recipe, like the ingredients or instructions, come in the form of lists where we need to find multiple elements with the same attributes. We can use `self.soup.find_all()` for this. `find_all` uses the same arguments as `find`, it just returns a list of `bs4.element.Tag` objects with all the matching elements.
+
+Using the same site as above, we can find all the ingredients like so
+
+```python
+ingredient_tags = self.soup.find_all("li", class_="wprm-recipe-ingredient")
+ingredients_text = [normalize_string(tag.get_text()) for tag in ingredient_tags]
+"""
+[
+ '2 (20-ounce // 565g) cans of jackfruit (in water or brine)*',
+ '1 tablespoon coconut oil (or oil of choice)',
+ '1 1/2 teaspoons cumin seeds',
+ '1 1/2 teaspoons black mustard seeds (can substitute brown mustard seeds)',
+ '1 large yellow onion, diced',
+ ...
+]
+"""
+```
+
+The Beautiful Soup documentation for `find_all` is [here](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#find-all).
+
+## Using CSS selectors
+
+If you are already familiar with CSS selectors, then you can `select()` to achieve the same result as `find_all()`, or `select_one()` to achieve the same result as `find`.
+
+```python
+ingredient_tag = self.soup.select("li.wprm-recipe-ingredient") # Match all li elements with wprm-recipe-ingredient class
+```
+
+The Beautiful Soup documentation for `select` is [here](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#css-selectors-through-the-css-property). MDN has a guide on CSS selectors [here](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Selectors).
+
+## Finding elements using a partial attribute
```

Comment:
The HTML scraping guide is looking pretty good to me already :)

I'd suggest calling this section "Finding elements using string pattern matching"

Reply:
...or attribute pattern matching, I suppose (I see your point now that the matching is specific to HTML attributes)